Move IpAddr, SocketAddr and V4+V6 related types to `core`

faern opened the pull request:
Implements RFC https://github.com/rust-lang/rfcs/pull/2832. The RFC has completed FCP with disposition merge, but is not yet merged.

Moves IP types to `core` as specified in the RFC.

The full list of moved types is: `IpAddr`, `Ipv4Addr`, `Ipv6Addr`, `SocketAddr`, `SocketAddrV4`, `SocketAddrV6`, `Ipv6MulticastScope` and `AddrParseError`.

Doing this move was one of the main driving arguments behind #78802.

faern committed:
Move IpAddr and SocketAddr to core

faern commented:
@rustbot label +T-libs-api -T-libs

faern commented:
Should moves like this mark the new types in `core::net` an unstable? I found no way of doing that. Maybe I made a mistake.

Also, what to do with all the existing stability attributes, such as:
```rust
#[stable(feature = "ip_shared", since = "1.12.0")]
pub const fn is_loopback(&self) -> bool {
```
Do we leave them as is in moves like this? I have not yet found another PR doing a similar move.

slanterns commented:
> Should moves like this mark the new types in `core::net` an unstable? I found no way of doing that. Maybe I made a mistake.

I recall a similar attempt which moves `Error` to libcore. Maybe you can refer to https://github.com/rust-lang/rust/pull/95956 and https://github.com/rust-lang/rust/pull/99917/.

thomcc commented:
I think this is blocked on the RFC at the moment.

bors commented:
:umbrella: The latest upstream changes (presumably #104428) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

pinkforest commented:
Has anyone done testing about code sizes in microcontroller / wasm32-u-u context / between code size sensitive targets ? I think tests should be essential to "bench" / measure these kind of things if core is to be added with this stuff that realistically should be standard library stuff - this would alleviate concerns that binary sizes are kept in check.

Use-cases are: 
1/ `std` wasm32-unknown-unknown
2/ `no_std` + detached standard `alloc` (`[dlmalloc](https://crates.io/crates/dlmalloc)` crate)
3/ `no_std` + [lol_alloc](https://crates.io/crates/lol_alloc) 300 byte allocator w/ correct type (wee_alloc leaks)
4/ `no_std` + [heapless](https://crates.io/crates/heapless) (crate)
5/ `no_std` + no heap at all (e.g. no `String`/`Vec`)

e.g. w/ `cdylib` context

```
[profile.release]
codegen-units = 1
strip = "symbols"
debug = false
lto = true
opt-level = "z"
panic = 'abort'
```

pinkforest commented:
>  what is being measured. 

Binary sizes - There was some other related effort out there somewhere else and might be worthwhile to check if anything was done re: code sizes testing in general and if this could be incorporated somehow.

> Embedded no_std binaries are entirely statically linked, and unused symbols are stripped.

Yes that is true and correct. However it could be helpful to come up with a test that stays around and ensures this is correct in case of core::net e.g. for wasm32-u-u that the symbol.

Question may / probably is how to test it if this is done and what are the downsides of such test if any ?

newAM commented:
> Binary sizes - There was some other related effort out there somewhere else and might be worthwhile to check if anything was done re: in general and if this could be incorporated somehow there.

I checked binary sizes of the [cortex-m-quickstart](https://github.com/rust-embedded/cortex-m-quickstart) built for the `thumbv7em-none-eabi` target with and without this patch (with slight modifications to make the patch work with the 1.65.0 release), section sizes and symbol sizes are identical for both binaries.  Rough work is here: [core-net-benchmark](https://github.com/newAM/core-net-benchmark/commit/180b33715d2cd6a3fceeccca3dbaaf54ed49f244).

> Yes that is true and correct. However it could be helpful to come up with a test that stays around and ensures this is correct in case of core::net e.g. for wasm32-u-u that the symbol was stripped for completeness.

Is this not covered by existing dead code tests, such as `src/test/codegen/link-dead-code.rs`?

faern commented:
Rebased and updated. Since the RFC seems to be close to being merged.

Changed the std re-exports into type aliases so they can remain stable while the core versions are unstable (under the new feature `ip_in_core`). However, for `Ipv6MulticastScope` I kept it a `pub use`. Otherwise I ran into [this error](https://github.com/rust-lang/rust/actions/runs/3862619716/jobs/6584302375#step:26:2270). And I don't think it should matter, because it's still an `unstable` type and I suppose it will be stabilized in `core` and `std` at the same time anyway. So no need to be able to have them separate?

Not going to touch anything regarding testing/benchmarking the size of `core` in this PR unless someone from the libs team tell me to. That concern seems unrelated to this PR and seems to be a concern regarding linker optimizations in general, not anything specific to IP types at all.

The CI failure is something I can reproduce locally but have not yet figured out how to solve :thinking:

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3862782115/jobs/6584559377) [(plain)](https://github.com/rust-lang/rust/commit/880a99f74221d2ce78f9cebab67018c699d81406/checks/10500487936/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

43    |                                            |
44    |                                            required by a bound introduced by this call
45    |
-    = note: multiple `impl`s satisfying `u32: From<_>` found in the following crates: `core`, `std`:
-            - impl From<Ipv4Addr> for u32;
+    = note: multiple `impl`s satisfying `u32: From<_>` found in the `core` crate:
48            - impl From<NonZeroU32> for u32;
49            - impl From<bool> for u32;
50            - impl From<char> for u32;

+            - impl From<core::net::ip_addr::Ipv4Addr> for u32;
51            - impl From<u16> for u32;
52            - impl From<u8> for u32;
53            - impl<T> From<!> for T;

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/issue-77982/issue-77982.stderr
To update references, rerun the tests and pass the `--bless` flag
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args traits/issue-77982.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/traits/issue-77982.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/issue-77982" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/issue-77982/auxiliary"
stdout: none
--- stderr -------------------------------
  --> /checkout/src/test/ui/traits/issue-77982.rs:8:10
   |
   |
LL |     opts.get(opt.as_ref()); //~ ERROR type annotations needed
   |          ^^^ ------------ type must be known at this point
   |          |
   |          cannot infer type of the type parameter `Q` declared on the associated function `get`
   |
   = note: multiple `impl`s satisfying `String: Borrow<_>` found in the following crates: `alloc`, `core`:
           - impl Borrow<str> for String;
           - impl<T> Borrow<T> for T
             where T: ?Sized;
note: required by a bound in `HashMap::<K, V, S>::get`
  --> /rustc/FAKE_PREFIX/library/std/src/collections/hash/map.rs:876:5
help: consider specifying the generic argument
   |
LL |     opts.get::<Q>(opt.as_ref()); //~ ERROR type annotations needed

error[E0283]: type annotations needed
  --> /checkout/src/test/ui/traits/issue-77982.rs:8:10
   |
   |
LL |     opts.get(opt.as_ref()); //~ ERROR type annotations needed
   |          ^^^     ------ type must be known at this point
   |          |
   |          cannot infer type of the type parameter `Q` declared on the associated function `get`
   |
   = note: multiple `impl`s satisfying `String: AsRef<_>` found in the following crates: `alloc`, `std`:
           - impl AsRef<OsStr> for String;
           - impl AsRef<Path> for String;
           - impl AsRef<[u8]> for String;
           - impl AsRef<str> for String;
help: consider specifying the generic argument
   |
LL |     opts.get::<Q>(opt.as_ref()); //~ ERROR type annotations needed

error[E0283]: type annotations needed
  --> /checkout/src/test/ui/traits/issue-77982.rs:13:59
   |
   |
LL |     let ips: Vec<_> = (0..100_000).map(|_| u32::from(0u32.into())).collect();
   |                                            |
   |                                            required by a bound introduced by this call
   |
   |
   = note: multiple `impl`s satisfying `u32: From<_>` found in the `core` crate:
           - impl From<NonZeroU32> for u32;
           - impl From<bool> for u32;
           - impl From<char> for u32;
           - impl From<core::net::ip_addr::Ipv4Addr> for u32;
           - impl From<u16> for u32;
           - impl From<u8> for u32;
           - impl<T> From<!> for T;
           - impl<T> From<T> for T;
   |
   |
LL |     let ips: Vec<_> = (0..100_000).map(|_| u32::from(<u32 as Into<T>>::into(0u32))).collect();

error[E0283]: type annotations needed for `Box<T>`
  --> /checkout/src/test/ui/traits/issue-77982.rs:36:9
   |
   |
LL |     let _ = ().foo(); //~ ERROR type annotations needed
   |         ^      --- type must be known at this point
   |
note: multiple `impl`s satisfying `(): Foo<'_, _>` found
   |
   |
LL | impl Foo<'static, u32> for () {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
LL | impl<'a> Foo<'a, i16> for () {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider giving this pattern a type, where the type for type parameter `T` is specified
   |
LL |     let _: Box<T> = ().foo(); //~ ERROR type annotations needed

error[E0283]: type annotations needed for `Box<T>`
  --> /checkout/src/test/ui/traits/issue-77982.rs:40:9
   |
   |
LL |     let _ = (&()).bar(); //~ ERROR type annotations needed
   |         ^         --- type must be known at this point
   |
note: multiple `impl`s satisfying `&(): Bar<'_, _>` found
   |
   |
LL | impl<'a> Bar<'static, u32> for &'a () {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
LL | impl<'a> Bar<'a, i16> for &'a () {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider giving this pattern a type, where the type for type parameter `T` is specified
   |
LL |     let _: Box<T> = (&()).bar(); //~ ERROR type annotations needed

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0283`.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #107187) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

faern committed:
Adapt issue-77982.stderr to new rustc error

faern commented:
I fixed the merge conflict and the CI error that existed. The stderr of a compiler error changed slightly, so I just needed to adapt the test's expected output.

faern commented:
To fix [this](https://github.com/rust-lang/rust/actions/runs/4010412559/jobs/6886894326#step:26:3447) error it looks like we need to update the book. But that's a separate git module. And it's also temporary, while the types are unstable in core... Will the book changes be merged before this PR is approved?

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4010412559/jobs/6886894326) [(plain)](https://github.com/rust-lang/rust/commit/1d5f7c77a6e3b8cbd96ef29602437eeb900cf3c1/checks/10889713957/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:ac593985615ec2ede58e132d2e21d2b1cbd6127c)
Download action repository 'rust-lang/simpleinfra@master' (SHA:8c19d39c6d9d7e831f6e393b2a871216393a5761)
Complete job name: PR (x86_64-gnu-llvm-13, false, ubuntu-20.04-xl)
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-llvm-13
---
Rustbook (x86_64-unknown-linux-gnu) - edition-guide
Rustbook (x86_64-unknown-linux-gnu) - style-guide
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.23s
book/ch06-01-defining-an-enum.html:309: broken link - `std/net/enum.IpAddr.html`
book/print.html:5534: broken link - `std/net/enum.IpAddr.html`
number of HTML files scanned: 33469
number of HTML redirects found: 10249
number of links checked: 2965056
number of links ignored due to external: 99180
```

</details>

faern commented:
It only changes URL because the type will **temporarily** be a type alias instead of a re-export. And it's like that to be able to have the `core` versions unstable while the `std` versions (of course) remain stable. An alternative is to insta-stabilize the `core` versions of these types, that would simplify some things quite a bit.

If there are other ways to avoid the URL change I don't know. I'm not familiar with rustdoc in that way.

faern commented:
I have never moved things to `core` before. But it feels like this can't be the first time the problem occurs?

lukas-code commented:
I think the `core::net` module itself needs to be unstable and everything inside the module stable (with the same stability attributes as in `std`). Then it should be possible to reexport the items to `std` as stable, but have them unstable in `core`.

faern commented:
You are correct. It all worked out by marking the `core::net` module unstable. I could then have all types in it be stable and just re-exporting them in `std::net`. It's still not possible to use any type in `core::net` without opting in to the `ip_in_core` feature, so the types do not become directly stable even in `core`. I think this is a good way forward.

pitaj commented:
@rustbot label +A-no_std +A-net +S-waiting-on-review -S-waiting-on-author

faern commented:
I realized I had left a bunch of duplicate tests. A lot of tests from `library/std/src/net/ip_addr/tests.rs` that had been copied into `library/core/tests/net/ip_addr.rs` were still left in the original file. Now I removed that duplication, just keeping the single `to_socket_addr_socketaddr` tests that is still relevant for the `std` crate.

bors commented:
:umbrella: The latest upstream changes (presumably #106677) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joshtriplett commented on library/core/src/net/mod.rs:
```diff
@@ -0,0 +1,24 @@
+//! * [`SocketAddr`] represents socket addresses of either IPv4 or IPv6; [`SocketAddrV4`]
+//!   and [`SocketAddrV6`] are respectively IPv4 and IPv6 socket addresses
+
+#![unstable(feature = "ip_in_core", issue = "none")]
```
```suggestion
#![unstable(feature = "ip_in_core", issue = "108443")]
```

joshtriplett committed:
Add tracking issue

joshtriplett commented:
r=me with the merge conflicts fixed.

faern commented:
Merge conflicts fixed! Happy to see this moving :partying_face:

r=joshtriplett

faern commented:
r=@joshtriplett 

Not sure which format the bot wants

slanterns commented:
Should at bors first, like @bors r=joshtriplett

bors commented:
@slanterns: :key: Insufficient privileges: Not in reviewers

faern commented:
@bors r=joshtriplett

albertlarsan68 commented:
@bors r=joshtriplett

bors commented:
:pushpin: Commit 1291216ac986ccc1432bf0bf2db9545e5251dd24 has been approved by `joshtriplett`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joshtriplett 1291216ac986ccc1432bf0bf2db9545e5251dd24 -->
<!-- homu: {"type":"Approved","sha":"1291216ac986ccc1432bf0bf2db9545e5251dd24","approver":"joshtriplett","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

RalfJung commented on library/core/tests/lib.rs:
```diff
@@ -77,6 +78,9 @@
 #![feature(iterator_try_reduce)]
+#![feature(const_ip)]
+#![feature(const_ipv4)]
+#![feature(const_ipv6)]
```
I am very surprised that this does not need the `ip_in_core` feature? miri-test-libstd is [failing due to that feature missing somewhere](https://github.com/rust-lang/miri-test-libstd/actions/runs/4288908148/jobs/7471415401).

faern replied:
I guess because `core` itself does not use anything from `core::net` (which is the unstable module). Submodules of `core::net` uses parts of itself, but I guess if you are inside an unstable module you can freely use the stuff in it?

RalfJung replied:
This is a different crate though, it is the coretest crate, not the core crate.

faern replied:
Ah, sorry. I misread the path! Yeah, that's very strange. But I see you have a PR up for fixing it :+1:

RalfJung replied:
Yeah it fixes miri-test-libstd. But still I am worried that the fact that it builds here without the feature flag might be indicative of a more fundamental problem -- as mentioned elsewhere, historically reexport stability did not work properly.

RalfJung commented on library/core/tests/net/parser.rs:
```diff
@@ -1,6 +1,6 @@
-use crate::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, SocketAddrV6};
-use crate::str::FromStr;
+use core::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, SocketAddrV6};
+use core::str::FromStr;
```
I think these tests are dead code now. Unit tests are not run for libcore:

https://github.com/rust-lang/rust/blob/1291216ac986ccc1432bf0bf2db9545e5251dd24/library/core/Cargo.toml#L13-L15

RalfJung commented on library/std/src/net/ip_addr.rs:
```diff
@@ -2,2101 +2,40 @@
-pub struct Ipv4Addr {
-    octets: [u8; 4],
-}
+pub use core::net::{Ipv4Addr, Ipv6Addr};
```
FWIW, I was under the impression that `pub use` with different stability doesn't actually work -- that's why `intrinsics::transmute` has to be stable, so that it can be `pub use`d in `mem`. Has that been fixed?

See https://github.com/rust-lang/rust/issues/18517, which is closed as a duplicate of an issue that is still open.

RalfJung replied:
Quick testing indicates that [the feature gate works as expected](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d1b5f6651898f896d2e982409cc5e4d4) (except for the strange behavior in coretest).

Cc @petrochenkov your [statement here](https://github.com/rust-lang/rust/pull/93017#issuecomment-1019151425) sounds like reexport stability doesn't work properly, has that changed since last year?

klensy commented:
This reduced `std` crate size by 200kb under `x86_64-pc-windows-msvc`, but i don't understand why.
